test/models/erratum_test.rb

```diff
@@ -72,15 +72,15 @@ def test_applicable_to_hosts_dashboard
     end
 
     def test_applicable_to_hosts_dashboard_respects_filter
-      assert Erratum.applicable_to_hosts_dashboard(::Host.search_for("compute_resource = SOMENAME")).empty?
+      assert Erratum.applicable_to_hosts_dashboard(::Host.where(compute_resource: "SOMENAME")).empty?
       host = FactoryGirl.build(:host, :with_content, :with_subscription,
                                       :content_view => katello_content_views(:library_dev_view),
                                       :lifecycle_environment => katello_environments(:library),
                                       :compute_resource_id => compute_resources(:one).id)
       host.save
       host.content_facet.applicable_errata << @security
       host.save
-      refute Erratum.applicable_to_hosts_dashboard(::Host.search_for("compute_resource = #{compute_resources(:one).name}")).empty?
+      refute Erratum.applicable_to_hosts_dashboard(::Host.where(compute_resource: compute_resources(:one))).empty?
```

Comment:
why was the search_for not working?

Reply:
Here was the failure:

```
  1) Error:                                                                                                                                                                                                                                                                     
Katello::ErratumTest#test_applicable_to_hosts_dashboard_respects_filter:                                                                                                                                                                                                        
ArgumentError: Unsupported query object: {:compute_resource=>"bigcompute"}!                                                                                                                                                                                                         app/models/host.rb:15:in `method_missing'                                                                                                                                                                                                                                   
    /home/vagrant/katello/test/models/erratum_test.rb:83:in `test_applicable_to_hosts_dashboard_respects_filter'
```

Seems like an error coming from scoped search itself.

Reply:
I'm not sure that is the error you got, when i try it, i get:  
```
  1) Error:
Katello::ErratumTest#test_applicable_to_hosts_dashboard_respects_filter:
ActiveRecord::StatementInvalid: PG::UndefinedTable: ERROR:  missing FROM-clause entry for table "compute_resources"
LINE 1: ..." WHERE "hosts"."type" IN ('Host::Managed') AND (("compute_r...
                                                             ^
: SELECT COUNT(count_column) FROM (SELECT  1 AS count_column FROM "katello_errata" INNER JOIN "katello_content_facet_errata" ON "katello_content_facet_errata"."erratum_id" = "katello_errata"."id" INNER JOIN "katello_content_facets" ON "katello_content_facets"."id" = "katello_content_facet_errata"."content_facet_id" WHERE "katello_content_facets"."host_id" IN (SELECT "hosts"."id" FROM "hosts" WHERE "hosts"."type" IN ('Host::Managed') AND (("compute_resources"."name" = 'SOMENAME'))  ORDER BY "hosts"."name" ASC) LIMIT 6) subquery_for_count
    ./katello/test/models/erratum_test.rb:75:in `test_applicable_to_hosts_dashboard_respects_filter'
```

but if you perform the search on the console it works fine:
```
> Host.search_for("compute_resource = foobar").count
=> 0
```

I think there is some issue with passing that search (or maybe any search?) into `applicable_for_hosts`.  This search test was added explicitly to test a search scenario since you can search from the dashboard, so we need to figure out why its not working

Reply:
@jlsherrill maybe you could help me out here and also verify that you are seeing this too. It seems somehow related to the `compute_resource` column itself.

```rb
irb(main):036:0> h1 = Host.search_for("compute_resource = SOMENAME")
=> #<ActiveRecord::Relation []>
irb(main):037:0> h2 = Host.search_for("name = SOMENAME")
=> #<ActiveRecord::Relation []>
irb(main):038:0> Katello::Erratum.joins(:content_facets).where("#{Katello::Host::ContentFacet.table_name}.host_id" => h2)
=> #<ActiveRecord::Relation []>
irb(main):039:0> Katello::Erratum.joins(:content_facets).where("#{Katello::Host::ContentFacet.table_name}.host_id" => h1)
Hirb Error: PG::UndefinedTable: ERROR:  missing FROM-clause entry for table "compute_resources"
LINE 1: ..." WHERE "hosts"."type" IN ('Host::Managed') AND (("compute_r...
                                                             ^
: SELECT "katello_errata".* FROM "katello_errata" INNER JOIN "katello_content_facet_errata" ON "katello_content_facet_errata"."erratum_id" = "katello_errata"."id" INNER JOIN "katello_content_facets" ON "katello_content_facets"."id" = "katello_content_facet_errata"."content_facet_id" WHERE "katello_content_facets"."host_id" IN (SELECT "hosts"."id" FROM "hosts" WHERE "hosts"."type" IN ('Host::Managed') AND (("compute_resources"."name" = 'SOMENAME'))  ORDER BY "hosts"."name" ASC)
    /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activerecord-4.2.9/lib/active_record/connection_adapters/postgresql_adapter.rb:592:in `async_exec'
    /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activerecord-4.2.9/lib/active_record/connection_adapters/postgresql_adapter.rb:592:in `block in exec_no_cache'
    /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activerecord-4.2.9/lib/active_record/connection_adapters/abstract_adapter.rb:484:in `block in log'
    /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activesupport-4.2.9/lib/active_support/notifications/instrumenter.rb:20:in `instrument'
    /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activerecord-4.2.9/lib/active_record/connection_adapters/abstract_adapter.rb:478:in `log'
    /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activerecord-4.2.9/lib/active_record/connection_adapters/postgresql_adapter.rb:592:in `exec_no_cache'
    /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activerecord-4.2.9/lib/active_record/connection_adapters/postgresql_adapter.rb:584:in `execute_and_clear'
    /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activerecord-4.2.9/lib/active_record/connection_adapters/postgresql/database_statements.rb:160:in `exec_query'
    /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activerecord-4.2.9/lib/active_record/connection_adapters/abstract/database_statements.rb:356:in `select'
    /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activerecord-4.2.9/lib/active_record/connection_adapters/abstract/database_statements.rb:32:in `select_all'
    /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activerecord-4.2.9/lib/active_record/connection_adapters/abstract/query_cache.rb:70:in `select_all'
    /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activerecord-4.2.9/lib/active_record/querying.rb:39:in `find_by_sql'
    /home/vagrant/.rvm/gems/ruby-2.2.4/gems/bullet-5.6.1/lib/bullet/active_record42.rb:24:in `find_by_sql'
    /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activerecord-4.2.9/lib/active_record/relation.rb:639:in `exec_queries'
    /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activerecord-4.2.9/lib/active_record/relation.rb:515:in `load'
    /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activerecord-4.2.9/lib/active_record/relation.rb:243:in `to_a'
    /home/vagrant/.rvm/gems/ruby-2.2.4/gems/bullet-5.6.1/lib/bullet/active_record42.rb:67:in `to_a'
    /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activerecord-4.2.9/lib/active_record/relation/delegation.rb:46:in `to_ary'
    /home/vagrant/.rvm/gems/ruby-2.2.4/gems/hirb-0.7.3/lib/hirb/formatter.rb:88:in `Array'
    /home/vagrant/.rvm/gems/ruby-2.2.4/gems/hirb-0.7.3/lib/hirb/formatter.rb:88:in `determine_output_class'
    /home/vagrant/.rvm/gems/ruby-2.2.4/gems/hirb-0.7.3/lib/hirb/formatter.rb:53:in `format_output'
    /home/vagrant/.rvm/gems/ruby-2.2.4/gems/hirb-0.7.3/lib/hirb/view.rb:205:in `render_output'
    /home/vagrant/.rvm/gems/ruby-2.2.4/gems/hirb-0.7.3/lib/hirb/view.rb:123:in `view_output'
    /home/vagrant/.rvm/gems/ruby-2.2.4/gems/hirb-0.7.3/lib/hirb/view.rb:201:in `view_or_page_output'
    /home/vagrant/.rvm/gems/ruby-2.2.4/gems/hirb-0.7.3/lib/hirb/view.rb:187:in `output_value'
    /home/vagrant/.rvm/rubies/ruby-2.2.4/lib/ruby/2.2.0/irb.rb:490:in `block (2 levels) in eval_input'
    /home/vagrant/.rvm/rubies/ruby-2.2.4/lib/ruby/2.2.0/irb.rb:623:in `signal_status'
    /home/vagrant/.rvm/rubies/ruby-2.2.4/lib/ruby/2.2.0/irb.rb:486:in `block in eval_input'
    /home/vagrant/.rvm/rubies/ruby-2.2.4/lib/ruby/2.2.0/irb/ruby-lex.rb:245:in `block (2 levels) in each_top_level_statement'
    /home/vagrant/.rvm/rubies/ruby-2.2.4/lib/ruby/2.2.0/irb/ruby-lex.rb:231:in `loop'
    /home/vagrant/.rvm/rubies/ruby-2.2.4/lib/ruby/2.2.0/irb/ruby-lex.rb:231:in `block in each_top_level_statement'
    /home/vagrant/.rvm/rubies/ruby-2.2.4/lib/ruby/2.2.0/irb/ruby-lex.rb:230:in `catch'
    /home/vagrant/.rvm/rubies/ruby-2.2.4/lib/ruby/2.2.0/irb/ruby-lex.rb:230:in `each_top_level_statement'
    /home/vagrant/.rvm/rubies/ruby-2.2.4/lib/ruby/2.2.0/irb.rb:485:in `eval_input'
    /home/vagrant/.rvm/rubies/ruby-2.2.4/lib/ruby/2.2.0/irb.rb:395:in `block in start'
    /home/vagrant/.rvm/rubies/ruby-2.2.4/lib/ruby/2.2.0/irb.rb:394:in `catch'
    /home/vagrant/.rvm/rubies/ruby-2.2.4/lib/ruby/2.2.0/irb.rb:394:in `start'
    /home/vagrant/.rvm/gems/ruby-2.2.4/gems/railties-4.2.9/lib/rails/commands/console.rb:110:in `start'
    /home/vagrant/.rvm/gems/ruby-2.2.4/gems/railties-4.2.9/lib/rails/commands/console.rb:9:in `start'
    /home/vagrant/.rvm/gems/ruby-2.2.4/gems/railties-4.2.9/lib/rails/commands/commands_tasks.rb:68:in `console'
    /home/vagrant/.rvm/gems/ruby-2.2.4/gems/railties-4.2.9/lib/rails/commands/commands_tasks.rb:39:in `run_command!'
    /home/vagrant/.rvm/gems/ruby-2.2.4/gems/railties-4.2.9/lib/rails/commands.rb:17:in `<top (required)>'
    /home/vagrant/.rvm/gems/ruby-2.2.4/gems/polyglot-0.3.5/lib/polyglot.rb:65:in `require'
    /home/vagrant/.rvm/gems/ruby-2.2.4/gems/polyglot-0.3.5/lib/polyglot.rb:65:in `require'
    /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activesupport-4.2.9/lib/active_support/dependencies.rb:274:in `block in require'
    /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activesupport-4.2.9/lib/active_support/dependencies.rb:240:in `load_dependency'
    /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activesupport-4.2.9/lib/active_support/dependencies.rb:274:in `require'
    /home/vagrant/foreman/bin/rails:9:in `<top (required)>'
    /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activesupport-4.2.9/lib/active_support/dependencies.rb:268:in `load'
    /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activesupport-4.2.9/lib/active_support/dependencies.rb:268:in `block in load'
    /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activesupport-4.2.9/lib/active_support/dependencies.rb:240:in `load_dependency'
    /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activesupport-4.2.9/lib/active_support/dependencies.rb:268:in `load'
    /home/vagrant/.rvm/rubies/ruby-2.2.4/lib/ruby/site_ruby/2.2.0/rubygems/core_ext/kernel_require.rb:55:in `require'
    /home/vagrant/.rvm/rubies/ruby-2.2.4/lib/ruby/site_ruby/2.2.0/rubygems/core_ext/kernel_require.rb:55:in `require'
    -e:1:in `<main>'
```

Reply:
Ah, so in the comment above, `h1` is not getting fully expressed in sql:

```rb
irb(main):043:0> Katello::Erratum.joins(:content_facets).where("#{Katello::Host::ContentFacet.table_name}.host_id" => h1)
=> # same error as above
irb(main):043:0> Katello::Erratum.joins(:content_facets).where("#{Katello::Host::ContentFacet.table_name}.host_id" => h1.to_sql)
=> #<ActiveRecord::Relation []>

irb(main):044:0> puts Katello::Erratum.joins(:content_facets).where("#{Katello::Host::ContentFacet.table_name}.host_id" => h1).to_sql
SELECT "katello_errata".* FROM "katello_errata" INNER JOIN "katello_content_facet_errata" ON "katello_content_facet_errata"."erratum_id" = "katello_errata"."id" INNER JOIN "katello_content_facets" ON "katello_content_facets"."id" = "katello_content_facet_errata"."content_facet_id" WHERE "katello_content_facets"."host_id" IN (SELECT "hosts"."id" FROM "hosts" WHERE "hosts"."type" IN ('Host::Managed') AND (("compute_resources"."name" = 'SOMENAME'))  ORDER BY "hosts"."name" ASC)
=> nil
irb(main):046:0> puts h1.to_sql
SELECT "hosts"."id" AS t0_r0, "hosts"."name" AS t0_r1, "hosts"."last_compile" AS t0_r2, "hosts"."last_report" AS t0_r3, "hosts"."updated_at" AS t0_r4, "hosts"."created_at" AS t0_r5, "hosts"."root_pass" AS t0_r6, "hosts"."architecture_id" AS t0_r7, "hosts"."operatingsystem_id" AS t0_r8, "hosts"."environment_id" AS t0_r9, "hosts"."ptable_id" AS t0_r10, "hosts"."medium_id" AS t0_r11, "hosts"."build" AS t0_r12, "hosts"."comment" AS t0_r13, "hosts"."disk" AS t0_r14, "hosts"."installed_at" AS t0_r15, "hosts"."model_id" AS t0_r16, "hosts"."hostgroup_id" AS t0_r17, "hosts"."owner_id" AS t0_r18, "hosts"."owner_type" AS t0_r19, "hosts"."enabled" AS t0_r20, "hosts"."puppet_ca_proxy_id" AS t0_r21, "hosts"."managed" AS t0_r22, "hosts"."use_image" AS t0_r23, "hosts"."image_file" AS t0_r24, "hosts"."uuid" AS t0_r25, "hosts"."compute_resource_id" AS t0_r26, "hosts"."puppet_proxy_id" AS t0_r27, "hosts"."certname" AS t0_r28, "hosts"."image_id" AS t0_r29, "hosts"."organization_id" AS t0_r30, "hosts"."location_id" AS t0_r31, "hosts"."type" AS t0_r32, "hosts"."otp" AS t0_r33, "hosts"."realm_id" AS t0_r34, "hosts"."compute_profile_id" AS t0_r35, "hosts"."provision_method" AS t0_r36, "hosts"."grub_pass" AS t0_r37, "hosts"."content_view_id" AS t0_r38, "hosts"."lifecycle_environment_id" AS t0_r39, "hosts"."global_status" AS t0_r40, "hosts"."lookup_value_matcher" AS t0_r41, "hosts"."pxe_loader" AS t0_r42, "compute_resources"."id" AS t1_r0, "compute_resources"."name" AS t1_r1, "compute_resources"."description" AS t1_r2, "compute_resources"."url" AS t1_r3, "compute_resources"."user" AS t1_r4, "compute_resources"."password" AS t1_r5, "compute_resources"."uuid" AS t1_r6, "compute_resources"."type" AS t1_r7, "compute_resources"."created_at" AS t1_r8, "compute_resources"."updated_at" AS t1_r9, "compute_resources"."attrs" AS t1_r10, "compute_resources"."email" AS t1_r11, "compute_resources"."caching_enabled" AS t1_r12 FROM "hosts" LEFT OUTER JOIN "compute_resources" ON "compute_resources"."id" = "hosts"."compute_resource_id" WHERE "hosts"."type" IN ('Host::Managed') AND (("compute_resources"."name" = 'SOMENAME'))  ORDER BY "hosts"."name" ASC
=> nil
```

So just using `h1.to_sql` fixes the issue.